### ISSUE TITLE
Question: Qwen 3.8 27B on Mac with Lora

### ISSUE THREAD

**Issue by Sir-Lysander** [MakazhanAlpamys/Soup#472]
Hello, I have a MacBook Pro M4 Max with 128GB of unified RAM. I was wondering if the Qwen 3.8 27B model is currently supported on my Mac for distillation using Lora?

**Comment by MakazhanAlpamys** [comment]
Hello — thanks for asking before burning a run on it. Short answer: **LoRA SFT yes, distillation no**, and the "no" is a real limitation rather than a missing model.

I checked the code rather than answering from memory, so here is exactly what holds today.

## 1. The model name

There is no **Qwen 3.8** in Soup's catalog, and as far as I know Alibaba has not released one. What exists here is:

```
Qwen/Qwen3.5-27B      -> recipe  qwen3.5-27b-sft
Qwen/Qwen3.6-27B      -> recipe  qwen3.6-27b-sft
```

If you meant one of those, read on. If 3.8 is something you have seen released, tell me and I will look — the catalog is hand-maintained and it would not be the first time we were behind.

## 2. Distillation on Apple Silicon is not supported, and it fails loudly

The MLX backend registers exactly three tasks — `sft`, `dpo`, `grpo`. `distill` is not one of them, and the config is refused at parse time rather than failing minutes in:

```
$ soup train -c soup.yaml     # task: distill, backend: mlx
REFUSED: task='distill' is not supported on backend=mlx
```

At least you get that in the first second and not after loading 27B of weights.

Two workarounds, both with honest caveats:

- **`backend: transformers` on MPS.** Nothing blocks it, and 128 GB unified is genuinely a lot — but distillation holds a **student and a frozen teacher simultaneously**, so a 27B student plus any serious teacher is a very different memory story from SFT. I have no Apple Silicon here (this box is a 4 GB RTX 3050), so I am not going to tell you it works. If you try it, I would genuinely like to know what happens.
- **Two-step instead**: generate from the teacher with `soup data best-of-n` or `soup data gen-magpie`, then plain LoRA SFT on the result via MLX. Not the same thing as logit-level KD, but it runs on the supported path.

## 3. What does work on your machine

LoRA SFT via MLX. The 27B recipe is written for a CUDA/DeepSpeed box, so switch the backend and drop the bitsandbytes quantization, which is CUDA-only:

```
soup recipes show qwen3.5-27b-sft
```

then set `backend: mlx` and `quantization: none` — or start from `qwen3-8b-sft-mlx`, which is already MLX-shaped, and swap the base.

## 4. One open issue you should know about before you start

**#394** — `pip install "soup-cli[mlx]"` alone may not train, because the MLX path imports `datasets` and TRL before reaching the MLX branch. It is open and **unreproduced** — nobody on the team has an Apple Silicon machine, which is precisely why it is still open.

So: if `soup train` dies on an import before any training starts, that is #394 and not something you did wrong. **A reproduction from a real M4 Max would be more valuable to us than most bug reports** — please comment there if you hit it.

Related and also open: **#434**, where layer streaming's pinned host store lands on `mps`. Not relevant to LoRA SFT, but it is the second Apple Silicon item we cannot test.

Leaving this issue open until you tell me whether the MLX SFT path actually gets you going.

**Comment by Sir-Lysander** [comment]
Thanks for your reply! I'll see if I can put your suggestion into practice. Since I have a ChatGPT subscription, maybe I could try to solve some issues using Codex—that might help Mac users get the most out of Soup. By the way, Alibaba recently launched its new model, Qwen3.8-27B, which continues to deliver exceptional performance given its size. Here’s the link to HuggingFace: https://huggingface.co/Qwen/Qwen3.8-27B

**Comment by MakazhanAlpamys** [comment]
Correction to my earlier answer: **I was wrong about the model.** `Qwen/Qwen3.8-27B` exists — created 2026-08-05, 2,090,699 downloads, 12,019 likes. I said "as far as I know Alibaba has not released one", and that was simply out of date. Thanks for filing #477 with the link rather than letting it stand.

What I got right stands, and one part now matters more than it did:

```
pipeline_tag  : image-text-to-text
model_type    : qwen3_5
architectures : ['Qwen3_5ForConditionalGeneration']
```

It is a **vision-language model**, not a text model. So for your original question — distillation with LoRA on the M4 Max — the answer is unchanged but for an additional reason: `task='distill'` is still refused on `backend='mlx'` at parse time (MLX registers only `sft`, `dpo`, `grpo`), and on top of that a VLM would need `modality: vision`, which narrows the supported paths further.

If you want to fine-tune it rather than distil into it, that is a different and more answerable question — say so and I will work out what actually runs on 128 GB unified.

Continuing the catalog side on #477.

**Comment by Amix29** [comment]
Thanks for clarifying. My goal is text-only LoRA SFT, not online or logit-level knowledge distillation.

I already have a question/answer dataset whose answers were generated by a stronger cloud model. The dataset contains no images and no multimodal examples. I want to fine-tune `Qwen/Qwen3.8-27B` on those text-only examples using my M4 Max with 128 GB of unified memory.

Operationally, this is offline response distillation implemented as ordinary supervised fine-tuning. I therefore want to use `task: sft` with `modality: text`.

The remaining question is whether Soup can safely use the language component of this `Qwen3_5ForConditionalGeneration` checkpoint, and whether the correct backend is MLX or Transformers/MPS.

**Comment by Amix29** [comment]
Completed the end-to-end Apple Silicon smoke test for `Qwen/Qwen3.8-27B` in BF16 with a high-capacity LoRA (`r=128`, `alpha=256`) on an Apple M4 Max with 128 GB unified memory.

This was deliberately a functional smoke test, not a quality benchmark: only 128 cleaned training examples and one epoch were used. The frozen test contained 445 held-out Tsez interlinear-glossing examples, and the same deterministic generation protocol was used before and after training.

## Result

The complete path worked: load -> train -> save -> reload -> generate.

| Metric | Base | LoRA | Change |
|---|---:|---:|---:|
| Exact match | 0/445 | 3/445 | +3 |
| Token edit similarity | 0.0500 | 0.2379 | +0.1879 (4.76x) |
| Gloss error rate | 1.0033 | 0.8117 | -19.1% |
| Token-count match | 44.04% | 82.25% | +38.20 pp |
| Outputs hitting 192-token cap | 40/445 | 4/445 | -90% |
| Output throughput | 5.70 tok/s | 5.34 tok/s | -6.35% |

Pairwise edit similarity improved on 352 examples, regressed on 10, and tied on 83. All 445 base evaluations and all 445 LoRA evaluations completed without an inference error. The adapter contained 211,812,352 trainable parameters (about 0.76%), saved successfully, reloaded as `PeftModelForCausalLM`, and all adapter tensors were finite.

The low absolute exact-match score is expected for 128 examples / one epoch and may also reflect dataset difficulty and hyperparameters. Model quality was not the main claim being tested. The useful result is that Soup genuinely trained and reloaded Qwen3.8-27B on this Mac, and the frozen test moved strongly in the expected direction.

## Performance and resource notes

- Trainer loop: 128 steps in 1,016 seconds (16m56s).
- Loss: 4.3004 on the first step, 1.3589 on the last; mean 1.3184.
- Base frozen evaluation: 89.0 minutes.
- LoRA frozen evaluation: 76.2 minutes because it produced far fewer runaway sequences, despite slightly lower per-token throughput.
- Training peak process RSS: about 58.2 GiB.
- Training peak sampled MPS driver allocation: about 24.5 GiB.
- Inference MPS allocation remained stable across both 445-example runs.

## Findings

1. The native Qwen3.8 template currently crashes `train_on_responses_only` on the system-only cumulative prefix. A generic ChatML override was required.
2. The final live panel replaces the last valid loss with `0.0000` when the Trainer emits its summary-only log.
3. The reproducibility receipt reports `gpu_models: []` on MPS.
4. The saved LoRA tensors were FP32 even though the base model and MPS run were BF16. This is PEFT's stability-oriented default, so I am treating it as a measurement/configuration question rather than immediately calling it a Soup bug. It is a plausible contributor to the measured 6.35% per-token overhead and deserves an isolated BF16-vs-FP32-vs-merged benchmark.

Targeted follow-ups:

- #540 — native Qwen3.8 response-only masking crash;
- #541 — final live panel resets the loss;
- #542 — MPS accelerator metadata missing from reproducibility receipts.

I am preparing small PRs with regression tests. No dataset rows, private paths or hardware identifiers are included here. The dataset is CC BY-NC 4.0, so this experiment is non-commercial.

**Comment by MakazhanAlpamys** [comment]
@Sir-Lysander — closing this as **answered, with a measured yes**. Your question was whether `Qwen/Qwen3.8-27B` works on an M4 Max with 128 GB for LoRA-based distillation. It does, and that is not my opinion — @Amix29 has the same machine and ran it end to end.

## The answer, from a real run

Load -> train -> save -> reload -> generate, all the way through, on an M4 Max / 128 GB unified memory. BF16 base, LoRA `r=128 / alpha=256`, 128 examples, one epoch, against 445 held-out examples with the same deterministic generation protocol before and after:

| metric | base | LoRA | change |
|---|---:|---:|---|
| exact match | 0/445 | 3/445 | +3 |
| token edit similarity | 0.0500 | 0.2379 | **4.76x** |
| gloss error rate | 1.0033 | 0.8117 | -19.1% |
| token-count match | 44.04% | 82.25% | +38.20pp |

```
training       128 steps / 1,016 s (16m56s)
loss           4.3004 first step -> 1.3589 last, mean 1.3184
peak RSS       ~58.2 GiB
peak MPS       ~24.5 GiB
```

That last line is the practically useful one for you: **~24.5 GiB of MPS allocation and ~58 GiB RSS**, so 128 GB has real headroom at this shape.

Two caveats stated plainly, because they were stated plainly in the run: this was a functional smoke test rather than a quality benchmark — 128 examples and one epoch — so the direction of movement is the result, not the magnitude. And on terminology: what you and @Amix29 both want is offline response distillation implemented as ordinary supervised fine-tuning, so `task: sft` with `modality: text` is the right configuration, not `task: distill`.

## It also found three real bugs, and all three are already fixed

The run did not just answer the question, it produced defects and they were filed and closed:

- **#540** — the native Qwen3.8 chat template crashed `train_on_responses_only` on the system-only cumulative prefix (a generic ChatML override was needed at the time)
- **#541** — the final live panel replaced the last valid loss with `0.0000`
- **#542** — the reproducibility receipt reported `gpu_models: []` on MPS

So if you run this today you should not need the ChatML workaround that run required.

One open observation worth knowing rather than acting on: the saved LoRA tensors came out FP32 while the base and the MPS run were BF16. That is PEFT's stability-oriented default, and it is a plausible contributor to a measured ~6.35% per-token overhead — deliberately **not** called a Soup bug without an isolated BF16-vs-FP32-vs-merged benchmark behind it.

## Credit

@Amix29 answered a hardware question I could not answer, on hardware I do not have, and then turned the byproducts into three fixed bugs. That is the most useful possible outcome for a `question` issue and it is worth naming rather than folding into a "closed" label.

@Sir-Lysander — you also flagged the model on 2026-08-21, days after release, which is where the whole thread started. Reopen this or open a new issue if your own run does not match, and please include the config; a second data point on different data would be genuinely useful.
